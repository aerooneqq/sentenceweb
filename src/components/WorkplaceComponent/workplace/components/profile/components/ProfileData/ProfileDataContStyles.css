
#profileDataOutterContainer{ 
  height: 90%;
  margin-right: 20px;
}

#profileDataContainer{
  width: 300px;
  height: 100%;

  background:  transparent;
  display: flex;
  flex-direction: column;

  background: #f9f9f9;
  border-style: none;
  border-radius: 10px;
}

#profileDataHeaderCont{
  display: flex;
}

#profileDataContentCont{ 
  height: 90%;
  width: 100%;

  background-color: transparent;

  display: flex;
  flex-grow: 1;
  flex-direction: column;

  margin-top: 20px;
}

.fadeInAnimation{ 
  animation: fadeIn 0.5s linear;
}

@keyframes fadeIn{ 
  from { 
    opacity: 0;
  }

  to { 
    opacity: 1;
  }
}

#profileDataContentCont .textBlock{ 
  margin-bottom: 15px;
}

#saveOrDiscardBlock{ 
  margin-bottom: 40px;
}

#discardCont{ 
  margin-top: 10px;
}

#profileDataLoaderContainer{ 
  width: 100%;
  height: 100%;

  display: flex;

  flex-direction: column;
  justify-content: center;
  align-items: center; 
}