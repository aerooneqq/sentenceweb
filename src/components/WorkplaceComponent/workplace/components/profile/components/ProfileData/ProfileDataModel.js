import React from "react";

//Profile data components.
import CareerData from "./profileDataAtoms/CareerData/CareerData";
import AuthProfileData from "./profileDataAtoms/AuthProfileData/AuthProfileData";
import NameData from "./profileDataAtoms/NameData/NameData";
import LocationData from "./profileDataAtoms/LocationData/LocationData";
import Friends from "./profileDataAtoms/subscribersAndSubscriptions/Friends/Friends";

/**
 * This class represents the model of a profile data. Profile data is devided in 5 sections: career data, authentication data,
 * name data, location data and frinds data (subscribers and subscriptions).
 */
export default class ProfileDataModel { 

    setUpdateUserFunction(func) { 
        this.updateUserFunction = func;

        return this;
    }

    configureModel() { 
        this.components = [ 
            {name: "Career data", component: <CareerData user = {this.user} changeUpdatedUser = {this.updateUserFunction} />},
            {name: "Authentication data", component:<AuthProfileData user = {this.user} changeUpdatedUser = {this.updateUserFunction} />},
            {name: "Name data", component:<NameData user = {this.user} changeUpdatedUser = {this.updateUserFunction} />},
            {name: "Location data", component:<LocationData user = {this.user} changeUpdatedUser = {this.updateUserFunction} />},
            {name: "Subscrivers and subscriptions", component:<Friends user = {this.user} changeUpdatedUser = {this.updateUserFunction} />},
        ];
    }

    /**
     * Updates the given property of the user.
     * @param {The name of the property which was updated} propertyName 
     * @param {The new value of the updated property} value 
     */
    changeUpdatedUser(propertyName, value) {
        this.user[propertyName] = value;
    }

    /**
     * Discards all changes in the _user object.
     */
    discardAllChanges() { 
        for (let key in this._user) { 
            this.user[key] = this.initialUser[key];
        }
    }

    /**
     * Returns the component which is under the given index.
     * @param {The index of the profile data} index 
     */
    getProfileDataContent(index) { 
        if (index < 0 || index >= this.components.length) { 
            return null;
        }

        return this.components[index].component;
    }

    /**
     * Gets the name of the profile data component under the given index.
     * @param {The index of the profile data} index 
     */
    getProfileDataName(index) { 
        if (index < 0 || index >= this.components.length) { 
            return null;
        }
        
        return this.components[index].name;
    }
}