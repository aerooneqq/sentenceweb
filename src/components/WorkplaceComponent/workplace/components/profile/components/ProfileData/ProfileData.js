import React, {Component, lazy, Suspense} from "react";

//Styles
import "./ProfileDataContStyles.css";

//Model
import ProfileDataModel from "./ProfileDataModel";


//Components
import ProfileDataLoader from "./ProfileDataLoader/ProfileDataLoader";
import ProfileDataHeader from "./profileDataAtoms/ProfileDataHeader/ProfileDataHeader";
import ProfileHeader from "../ProfileHeader/ProfileHeader";
import Loader from "../../../../../../loader/Loader";

const profileDataModel = new ProfileDataModel();

/**
 * The brief description of that class and components' dependencies:
 * 1) All profile text blocks are wrapped in a "textBlock" class, the style rules for this
 *    class are written in the ProfileDataContStyles.css
 * 2) All logic is placed in the ProfileDataModel class.
 * 
 * */
export default class ProfileData extends Component {
  constructor(props) {
    super(props);

    this.state = {
      isFirstLoad: true,
      isUpdating: true,
      userFriendsMode: "subscribers"
    };

    this.changeUpdatedUser = this.changeUpdatedUser.bind(this);
    this.discardChangesInProfileData = this.discardChangesInProfileData.bind(this);
    this.setUpdatingStatus = this.setUpdatingStatus.bind(this);
    this.changeUserFriendMode = this.changeUserFriendMode.bind(this);
    
    profileDataModel.setUpdateUserFunction(this.changeUpdatedUser).configureModel();
  }

  setUpdatingStatus(status) { 
    this.setState({ 
      isUpdating: status
    });
  }

  changeUserFriendMode(mode){ 
    this.setState({ 
      userFriendsMode: mode
    })
  }

  discardChangesInProfileData() { 
    profileDataModel.discardAllChanges();
  }

  changeUpdatedUser(propertyName, value) {
    profileDataModel.changeUpdatedUser(propertyName, value);
  }

  render() {
    if (this.state.isFirstLoad === true && this.props.user === null) { 
      return (
        <div id = "profileDataOutterContainer">
          <ProfileHeader header = "Settings"/>
          <div id = "profileDataContainer">
            <ProfileDataLoader />
          </div>  
        </div>
      )
    }
    else { 
      return (
          <div id = "profileDataOutterContainer">
            <ProfileHeader header = "Settings"/>
            <div id = "profileDataContainer">
                <div id = "profileDataHeaderCont">
                  <ProfileDataHeader dataName = {profileDataModel.getProfileDataName(this.props.currentUserDataIndex)}/>
                 </div>  
          
                <div className = "profileDataContentCont">
                    {profileDataModel.getProfileDataContent(this.props.currentUserDataIndex)}
                </div>
            </div>
          </div>
      )
    }
  }
}
