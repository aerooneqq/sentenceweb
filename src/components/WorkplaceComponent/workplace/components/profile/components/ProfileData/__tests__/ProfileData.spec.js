import React from "react";
import {shallow} from "enzyme";
import {shallowToJson} from "enzyme-to-json";
import ProfileData from "../ProfileData";

const props = { 
    user: {},
    currentUserDataIndex: -1
}

describe("ProfileData", () => { 
    it("should render correctly", () => { 
        expect(shallowToJson(shallow(<ProfileData {...props}/>))).toMatchSnapshot();
    });

    it("should have proper initial state", () => { 
        const profileData = shallow(<ProfileData {...props}/>);

        const properState = { 
            isFirstLoad: true,
            isUpdating: true,
            userFriendsMode: "subscribers"
        }

        for (let key in properState) { 
            expect(properState[key]).toBe(profileData.state(key));
        }
    });

    it("should change updating status when setUpdatingStatus is called", () => { 
        const profileData = shallow(<ProfileData {...props}/>);
        expect(profileData.state("isUpdating")).toBe(true);

        profileData.instance().setUpdatingStatus(false);
        expect(profileData.state("isUpdating")).toBe(false);
    })

    it("should change the userFriendsMode when changeUserFriendMode is called", () => { 
        const profileData = shallow(<ProfileData {...props}/>);
        expect(profileData.state("userFriendsMode")).toBe("subscribers");

        profileData.instance().changeUserFriendMode("subscriptions");
        expect(profileData.state("userFriendsMode")).toBe("subscriptions");
    });
});