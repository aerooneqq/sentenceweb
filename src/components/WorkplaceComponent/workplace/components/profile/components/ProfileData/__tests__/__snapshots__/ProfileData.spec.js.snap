// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProfileData should render correctly 1`] = `
<div
  id="profileDataOutterContainer"
>
  <ProfileHeader
    header="Settings"
  />
  <div
    id="profileDataContainer"
  >
    <div
      id="profileDataHeaderCont"
    >
      <ProfileDataHeader
        dataName={null}
      />
    </div>
    <div
      className="profileDataContentCont"
    />
  </div>
</div>
`;
